File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.cpp
#include <string>
#include <map>
#include <math.h>
#include <cctype>
#include <algorithm>
#define MUL 65536

using namespace std;

map<string, int> m1, m2;

int find_union_num() {
    int res = 0;
    for(auto it = m1.begin(); it != m1.end(); it++) {
        if(m2.find(it->first) == m2.end()) res += it->second;
        else res += max(it->second, m2[it->first]);
    }
    for(auto it = m2.begin(); it != m2.end(); it++) {
        if(m1.find(it->first) != m1.end()) continue;
        res += it->second;
    }
    return res;
}

int find_intersect_num() {
    int res = 0;
    for(auto it = m1.begin(); it != m1.end(); it++) {
        if(m2.find(it->first) == m2.end()) continue;
        res += min(it->second, m2[it->first]);
    }
    return res;
}

int solution(string str1, string str2) {
    transform(str1.begin(), str1.end(), str1.begin(), ::tolower);
    transform(str2.begin(), str2.end(), str2.begin(), ::tolower);
    
    for(int i = 0; i < str1.length() - 1; i++) {
        if(!isalpha(str1[i]) || !isalpha(str1[i + 1])) continue;
        string tmp = str1.substr(i, 2);
        m1[tmp]++;
    }
    for(int i = 0; i < str2.length() - 1; i++) {
        if(!isalpha(str2[i]) || !isalpha(str2[i + 1])) continue;
        string tmp = str2.substr(i, 2);
        m2[tmp]++;
    }
    
    if(m1.empty() && m2.empty()) return MUL;
    
    int union_num = find_union_num();
    int intersect_num = find_intersect_num();
    
    double jaccard = (double)intersect_num / (double)union_num;
    int answer = floor(jaccard * MUL);
    
    return answer;
}

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
from collections import Counter

def seperate(s):
    length = len(s)

    substr_list = list()    
    for idx in range(length - 1):
        substr = s[idx: idx + 2]
        
        if substr.isalpha():
            substr_list.append(substr)
            
    return Counter(substr_list)
        

def solution(str1, str2):
    str1 = str1.lower()
    str2 = str2.lower()
    
    substr1_counter = seperate(str1)
    substr2_counter = seperate(str2)
    
    if len(substr1_counter) == 0 and len(substr2_counter) == 0:
        answer = 1 * 65536
    else:
        intersection_score = 0
        union_score = 0

        for substr1 in substr1_counter:
            if substr1 in substr2_counter:
                intersection_score += min(substr1_counter[substr1], substr2_counter[substr1])
                union_score += max(substr1_counter[substr1], substr2_counter[substr1])
            else:
                union_score += substr1_counter[substr1]

        for substr2 in substr2_counter:
            if substr2 not in substr1_counter:
                union_score += substr2_counter[substr2]
        
        answer = int(intersection_score / union_score * 65536)
        
    return answer